tortoisebeacon/tortoise_beacon.go

```diff
@@ -193,76 +203,178 @@ func (tb *TortoiseBeacon) Close() {
+			log.Uint64("weight", weight))
+		return
 	}
 
-	if (epochID - 1).IsGenesis() {
-		return types.HexToHash32(genesisBeacon).Bytes(), nil
+	// checks if we have recorded this blockID before
+	if _, ok := entry.blocks[blockID]; ok {
+		tb.logger.With().Warning("block already reported beacon", epochID, blockID)
+		return
 	}
 
+	entry.blocks[blockID] = struct{}{}
+	entry.weight += weight
+	tb.logger.With().Debug("added beacon from block",
+		epochID,
+		blockID,
+		log.String("beacon", types.BytesToHash(beacon).ShortString()),
+		log.Uint64("weight", weight))
+}
+
+func (tb *TortoiseBeacon) findMostWeightedBeaconForEpoch(epoch types.EpochID) []byte {
 	tb.mu.RLock()
 	defer tb.mu.RUnlock()
-
-	beacon, ok := tb.beacons[epochID-1]
+	epochBeacons, ok := tb.beaconsFromBlocks[epoch]
 	if !ok {
-		tb.logger.With().Error("beacon is not calculated",
-			log.Uint32("target_epoch", uint32(epochID)),
-			log.Uint32("beacon_epoch", uint32(epochID-1)))
+		return nil
+	}
+	var mostWeight uint64
+	var beacon []byte
+	numBlocks := 0
+	for k, v := range epochBeacons {
+		if v.weight > mostWeight {
+			beacon = []byte(k)
+			mostWeight = v.weight
+		}
+		numBlocks += len(v.blocks)
+	}
+
+	logger := tb.logger.WithFields(epoch, log.Int("num_blocks", numBlocks))
+
+	if uint32(numBlocks) < tb.config.BeaconSyncNumBlocks {
+		logger.Debug("not enough blocks to determine beacon")
+		return nil
+	}
+
+	logger.With().Info("beacon determined for epoch",
+		log.String("beacon", types.BytesToHash(beacon).ShortString()),
+		log.Uint64("weight", mostWeight))
+	return beacon
+}
+
+// GetBeacon returns a Tortoise Beacon value as []byte for a certain epoch or an error if it doesn't exist.
+// TODO(nkryuchkov): consider not using (using DB instead)
+func (tb *TortoiseBeacon) GetBeacon(epochID types.EpochID) ([]byte, error) {
+	if epochID == 0 {
+		return nil, ErrZeroEpoch
+	}
 
-		return nil, ErrBeaconNotCalculated
+	beaconEpoch := epochID - 1
+	if beaconEpoch.IsGenesis() {
+		return types.HexToHash32(genesisBeacon).Bytes(), nil
 	}
 
-	return beacon.Bytes(), nil
+	beacon := tb.getBeacon(beaconEpoch)
+	if beacon != nil {
+		return beacon, nil
+	}
+
+	beacon = tb.getPersistedBeacon(beaconEpoch)
+	if beacon != nil {
+		return beacon, nil
+	}
+
+	tb.logger.With().Warning("beacon not available",
+		epochID,
+		log.Uint32("beacon_epoch", uint32(beaconEpoch)))
+	return nil, ErrBeaconNotCalculated
+}
+
+func (tb *TortoiseBeacon) getBeacon(epoch types.EpochID) []byte {
+	tb.mu.RLock()
+	defer tb.mu.RUnlock()
+	if beacon, ok := tb.beacons[epoch]; ok {
+		return beacon.Bytes()
+	}
+	return nil
 }
 
 func (tb *TortoiseBeacon) setBeacon(epoch types.EpochID, beacon types.Hash32) {
+	tb.persistBeacon(epoch, beacon)
 	tb.mu.Lock()
+	defer tb.mu.Unlock()
 	tb.beacons[epoch] = beacon
-	tb.mu.Unlock()
+}
+
+func (tb *TortoiseBeacon) persistBeacon(epoch types.EpochID, beacon types.Hash32) {
+	if tb.db == nil {
+		return
+	}
+	if err := tb.db.Put(epoch.ToBytes(), beacon.Bytes()); err != nil {
```

Comment:
if persistence is enabled this error should be checked, why to "optimize" this and make persistence best-effort?

Reply:
ok. i didn't think it's worth failing other component if there is error here. but if disk is failing, we have bigger issues. changed as you suggested